src/r1cs/sparse.rs

```diff
@@ -0,0 +1,228 @@
+//! # Sparse Matrices
+//!
+//! This module defines a custom implementation of CSR/CSC sparse matrices.
+//! Specifically, we implement sparse matrix / dense vector multiplication
+//! to compute the `A z`, `B z`, and `C z` in Nova.
+
+use abomonation::Abomonation;
+use abomonation_derive::Abomonation;
+use ff::PrimeField;
+use rayon::prelude::*;
+use serde::{Deserialize, Serialize};
+
+/// CSR format sparse matrix, We follow the names used by scipy.
+/// Detailed explanation here: https://stackoverflow.com/questions/52299420/scipy-csr-matrix-understand-indptr
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Abomonation)]
+#[abomonation_bounds(where <F as PrimeField>::Repr: Abomonation)]
+pub struct SparseMatrix<F: PrimeField> {
+  #[abomonate_with(Vec<F::Repr>)]
+  /// all non-zero values in the matrix
+  pub data: Vec<F>,
+  /// column indices
+  pub indices: Vec<usize>,
+  /// row information
+  pub indptr: Vec<usize>,
+  /// number of columns
+  pub cols: usize,
+}
+
+impl<F: PrimeField> SparseMatrix<F> {
+  /// 0x0 empty matrix
+  pub fn empty() -> Self {
+    SparseMatrix {
+      data: vec![],
+      indices: vec![],
+      indptr: vec![0],
+      cols: 0,
+    }
+  }
+
+  /// Construct from the COO representation; Vec<usize(row), usize(col), F>.
+  /// We implicitly sort the columns during construction.
+  pub fn new(matrix: &[(usize, usize, F)], rows: usize, cols: usize) -> Self {
+    let mut new_matrix = vec![vec![]; rows];
```

Comment:
I don't think this can be done, because the length of the rows is unknown. Of course we know that a capacity of `cols` is an upper bound, but then we would be allocating `O(rows * cols)` memory upfront. In practice, if the matrix is actually sparse then the rows will be small and the cost of resizing should be minimal.

Reply:
Oh, my apologies, my brain farted, and I forgot the semantics of the `rows` and `cols` arguments for a minute. my bad.

Reply:
AFAIU, `rows` (represented by `indptr`) is fixed to be the actual number of rows. I guess we could also do dynamic row sizing, but that would be non-conventional

Reply:
Fair enough, that makes sense